Invalid C code generation with `seq[array[...]]`

### Nim Version

Any of
```
Nim Compiler Version 2.2.4 [Linux: amd64]
Compiled at 2026-02-03
Copyright (c) 2006-2025 by Andreas Rumpf

git hash: f7145dd26efeeeb6eeae6fff649db244d81b212d
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.6 [Linux: amd64]
Compiled at 2026-02-03
Copyright (c) 2006-2025 by Andreas Rumpf

git hash: ab00c56904e3126ad826bb520d243513a139436a
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.7 [Linux: amd64]
Compiled at 2026-02-03
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: d48fc130c137f22d2e9d3c35eba215ed7684f186
active boot switches: -d:release
```

```
Nim Compiler Version 2.3.1 [Linux: amd64]
Compiled at 2026-02-03
Copyright (c) 2006-2026 by Andreas Rumpf

git hash: bfc27867187e28dd3b5f2a887450cfc2c465da98
active boot switches: -d:release
```

with either of
```
gcc (Debian 15.2.0-12) 15.2.0
Copyright (C) 2025 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
or
```
Debian clang version 21.1.8 (2)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-21/bin
```

### Description

```nim
type N = object
  b: seq[array[1'u, int]]
discard N(b: @[[0]])
```

Admittedly, it's not clear this should compile: one array has size `1'u` while the other has size `1`. But a C compilation error shouldn't happen regardless. It's also notable that many variations of this code do seem to treat `array[1, int]` and `array[1'u, int]` as at least implicitly convertible types.

### Current Output

```text
Hint: used config file '/tmp/nim23/config/nim.cfg' [Conf]
Hint: used config file '/tmp/nim23/config/config.nims' [Conf]
......................................................................
CC: system/exceptions.nim
CC: std/private/digitsutils.nim
CC: system/dollars.nim
CC: system.nim
CC: u.nim
/tmp/tmp.pgKbRZ2xbT/@mu.nim.c: In function ‘NimMainModule’:
/tmp/tmp.pgKbRZ2xbT/@mu.nim.c:143:9: error: incompatible types when assigning to type ‘tySequence__PaPqxATneMvaJgZdR29cujw’ from type ‘tySequence__iR7FNQIicfV9cQXx6NvdXYw’
  143 | T1_.b = colontmpD_;
      |         ^~~~~~~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -fno-strict-aliasing -pthread   -I/tmp/nim23/lib -I/tmp -o /tmp/tmp.pgKbRZ2xbT/@mu.nim.c.o /tmp/tmp.pgKbRZ2xbT/@mu.nim.c' failed with exit code: 1
```

### Expected Output

```text
No C compilation error
```

### Known Workarounds

_No response_

### Additional Information

_No response_

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -41,7 +41,7 @@ type
     CoType
     CoOwnerSig
     CoIgnoreRange
-    CoIgnoreRangeInarray
+    CoIgnoreRangeInArray
     CoConsiderOwned
     CoDistinct
     CoHashTypeInsideNode
@@ -221,14 +221,14 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     else:
       for a in t.kids: c.hashType a, flags+{CoIgnoreRange}, conf
   of tyRange:
-    if {CoIgnoreRange, CoIgnoreRangeInarray} * flags == {}:
+    if {CoIgnoreRange, CoIgnoreRangeInArray} * flags == {}:
       c &= char(t.kind)
       c.hashTree(t.n, {}, conf)
       c.hashType(t.elementType, flags, conf)
-    elif CoIgnoreRangeInarray in flags:
+    elif CoIgnoreRangeInArray in flags:
       # include only the length of the range (not its specific bounds)
       c &= char(t.kind)
-      let l = lastOrd(conf, t) - firstOrd(conf, t) + 1
+      let l = lengthOrd(conf, t)
       lowlevel l
     else:
       c.hashType(t.elementType, flags, conf)
@@ -261,7 +261,7 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     if tfVarargs in t.flags: c &= ".varargs"
   of tyArray:
     c &= char(t.kind)
-    c.hashType(t.indexType, flags-{CoIgnoreRange}+{CoIgnoreRangeInarray}, conf)
+    c.hashType(t.indexType, flags-{CoIgnoreRange}+{CoIgnoreRangeInArray}, conf)
     c.hashType(t.elementType, flags-{CoIgnoreRange}, conf)
   else:
     c &= char(t.kind)

--- a/compiler/sighashes.nim
+++ b/compiler/sighashes.nim
@@ -41,6 +41,7 @@ type
     CoType
     CoOwnerSig
     CoIgnoreRange
+    CoIgnoreRangeInarray
     CoConsiderOwned
     CoDistinct
     CoHashTypeInsideNode
@@ -220,10 +221,17 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     else:
       for a in t.kids: c.hashType a, flags+{CoIgnoreRange}, conf
   of tyRange:
-    if CoIgnoreRange notin flags:
+    if {CoIgnoreRange, CoIgnoreRangeInarray} * flags == {}:
       c &= char(t.kind)
       c.hashTree(t.n, {}, conf)
-    c.hashType(t.elementType, flags, conf)
+      c.hashType(t.elementType, flags, conf)
+    elif CoIgnoreRangeInarray in flags:
+      # include only the length of the range (not its specific bounds)
+      c &= char(t.kind)
+      let l = lastOrd(conf, t) - firstOrd(conf, t) + 1
+      lowlevel l
+    else:
+      c.hashType(t.elementType, flags, conf)
   of tyStatic:
     c &= char(t.kind)
     c.hashTree(t.n, {}, conf)
@@ -253,7 +261,7 @@ proc hashType(c: var MD5Context, t: PType; flags: set[ConsiderFlag]; conf: Confi
     if tfVarargs in t.flags: c &= ".varargs"
   of tyArray:
     c &= char(t.kind)
-    c.hashType(t.indexType, flags-{CoIgnoreRange}, conf)
+    c.hashType(t.indexType, flags-{CoIgnoreRange}+{CoIgnoreRangeInarray}, conf)
     c.hashType(t.elementType, flags-{CoIgnoreRange}, conf)
   else:
     c &= char(t.kind)

--- a/tests/array/tarray.nim
+++ b/tests/array/tarray.nim
@@ -605,3 +605,17 @@ block t18643:
   except IndexDefect:
     caught = true
   doAssert caught, "IndexDefect not caught!"
+
+
+# bug #25475
+block:
+  type N = object
+    b: seq[array[1'u, int]]
+  doAssert N(b: @[[0]]) == N(b: @[[0]])
+
+block:
+  var x: array[5..6, int] = [0, 1]
+  var y: array[1..2, int] = [0, 1]
+
+  doAssert x == y # compiles
+  doAssert @[x] == @[y]

